group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2038,6 +2075,22 @@ private CoordinatorResult<Void, CoordinatorRecord> classicGroupJoinToConsumerGro
             records
         );
 
+        // 4. Maybe downgrade the consumer group if the last member using the
+        // consumer protocol is replaced by the joining member.
+        String existingStaticMemberIdOrNull = group.staticMemberId(request.groupInstanceId());
+        boolean isValidDowngrade = existingStaticMemberIdOrNull != null &&
+            validateOnlineDowngradeWithReplacedMemberId(group, existingStaticMemberIdOrNull);
+        if (isValidDowngrade) {
+            records.addAll(
+                convertToClassicGroup(
```

Comment:
nit: I wonder if we could pass `records` to `convertToClassicGroup` now. What do you think?